// @ethereum-waffle/chai@3.1.1/dist/cjs/matchers/changeBalance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ethers_1 = require("ethers");
const utils_1 = require("./calledOnContract/utils");
const account_1 = require("./misc/account");
function supportChangeBalance(Assertion) {
    Assertion.addMethod('changeBalance', function (signer, balanceChange) {
        const subject = this._obj;
        const derivedPromise = Promise.all([
            getBalanceChange(subject, signer),
            account_1.getAddressOf(signer)
        ]).then(([actualChange, address]) => {
            this.assert(actualChange.eq(ethers_1.BigNumber.from(balanceChange)), `Expected "${address}" to change balance by ${balanceChange} wei, ` +
                `but it has changed by ${actualChange} wei`, `Expected "${address}" to not change balance by ${balanceChange} wei,`, balanceChange, actualChange);
        });
        this.then = derivedPromise.then.bind(derivedPromise);
        this.catch = derivedPromise.catch.bind(derivedPromise);
        this.promise = derivedPromise;
        return this;
    });
}
exports.supportChangeBalance = supportChangeBalance;
async function getBalanceChange(transaction, account) {
    if (typeof transaction === 'function') {
        return getBalanceChangeForTransactionCall(transaction, account);
    }
    else {
        return getBalanceChangeForTransactionResponse(transaction, account);
    }
}
async function getBalanceChangeForTransactionCall(transactionCall, account) {
    utils_1.ensure(account.provider !== undefined, TypeError, 'Provider not found');
    const balanceBefore = await account.provider.getBalance(account_1.getAddressOf(account));
    await transactionCall();
    const balanceAfter = await account.provider.getBalance(account_1.getAddressOf(account));
    return balanceAfter.sub(balanceBefore);
}
async function getBalanceChangeForTransactionResponse(transactionResponse, account) {
    utils_1.ensure(account.provider !== undefined, TypeError, 'Provider not found');
    const transactionBlockNumber = (await transactionResponse.wait()).blockNumber;
    const balanceAfter = await account.provider.getBalance(account_1.getAddressOf(account), transactionBlockNumber);
    const balanceBefore = await account.provider.getBalance(account_1.getAddressOf(account), transactionBlockNumber - 1);
    return balanceAfter.sub(balanceBefore);
}
